frontend/src/lib/ReportNode.ts

```diff
@@ -0,0 +1,509 @@
+        NO_REPORT_ID
+      ),
+      BlockNode.empty()
+    )
+  }
+
+  public constructor(main: BlockNode, sidebar: BlockNode) {
+    this.main = main
+    this.sidebar = sidebar
+  }
+
+  public applyDelta(
+    reportId: string,
+    delta: Delta,
+    metadata: ForwardMsgMetadata
+  ): ReportRoot {
+    const parentBlock = metadata.parentBlock as BlockPath
+    const parentBlockPath = parentBlock.path
+    const containerId = parentBlock.container
+    const { deltaId } = metadata
+
+    // Update Metrics
+    MetricsManager.current.incrementDeltaCounter(getContainerName(containerId))
+
+    // The full path to the ReportNode within the element tree.
+    // Used to find and update the element node specified by this Delta.
+    const deltaPath: number[] = [...parentBlockPath, deltaId]
+
+    switch (delta.type) {
+      case "newElement": {
+        const element = delta.newElement as Element
+        if (element.type != null) {
+          MetricsManager.current.incrementDeltaCounter(element.type)
+        }
+
+        // Track component instance name.
+        if (element.type === "componentInstance") {
+          const componentName = element.componentInstance?.componentName
+          if (componentName != null) {
+            MetricsManager.current.incrementCustomComponentCounter(
+              componentName
+            )
+          }
+        }
+
+        return this.addElement(
+          containerId,
+          deltaPath,
+          reportId,
+          element,
+          metadata
+        )
+      }
+
+      case "addBlock": {
+        MetricsManager.current.incrementDeltaCounter("new block")
+        return this.addBlock(
+          containerId,
+          deltaPath,
+          delta.addBlock as BlockProto,
+          reportId
+        )
+      }
+
+      case "addRows": {
+        MetricsManager.current.incrementDeltaCounter("add rows")
+        return this.addRows(
+          containerId,
+          deltaPath,
+          delta.addRows as NamedDataSet,
+          reportId
+        )
+      }
+
+      default:
+        throw new Error(`Unrecognized deltaType: '${delta.type}'`)
+    }
+  }
+
+  public clearStaleNodes(currentReportId: string): ReportRoot {
+    const main =
+      this.main.clearStaleNodes(currentReportId) || BlockNode.empty()
+    const sidebar =
+      this.sidebar.clearStaleNodes(currentReportId) || BlockNode.empty()
+
+    return new ReportRoot(main, sidebar)
+  }
+
+  /** Return a Set containing all Elements in the tree. */
+  public getElements(): Set<Element> {
+    const elements = new Set<Element>()
+    this.main.getElements(elements)
+    this.sidebar.getElements(elements)
+    return elements
+  }
+
+  private addElement(
+    containerId: number,
+    deltaPath: number[],
+    reportId: string,
+    element: Element,
+    metadata: ForwardMsgMetadata
+  ): ReportRoot {
+    const existingContainer = this.getContainer(containerId)
+
+    // TODO: do we care about this? Is it a meaningful optimization?
```

Comment:
I'm highly uncertain about this, but maybe this was for widget value persistence, to prevent React recreating the widget and resetting its value?

If that IS the case, the recent change to retrieve widget values out of WidgetStateManager may have obviated this... thoughts?

Reply:
I think it's just a potential optimization - if you have an element that's really complex, like a chart, then you don't want to rebuild that chart (even if you do re-render its Element class). 

That is, if you have a chart that remains consistent across two runs of the app, its `ElementNode` will always change (to get the updated `reportId`), but the element data itself won't. This optimization would allow us to reuse the original protobuf object in the updated `ElementNode` - but it would require us to deep-compare the protobufs, which we don't have a good way of doing (and which itself might be expensive enough that it throws out any potential re-render optimization).

So I think it's not worth having here - I'm not even sure it was worth having originally. I'm gonna remove the comment.